src/python/pants/backend/python/lint/autoflake/subsystem.py

```diff
@@ -0,0 +1,48 @@
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from typing import Tuple, cast
+
+from pants.backend.python.subsystems.python_tool_base import PythonToolBase
+from pants.backend.python.target_types import ConsoleScript
+from pants.option.custom_types import shell_str
+
+
+class Autoflake(PythonToolBase):
+    options_scope = "autoflake"
+    help = (
+        "The Python tool for removing unused/useless statements (imports, variables, duplicate keys, pass-es) "
+        "etc. (https://github.com/myint/autoflake)."
+    )
+
+    default_version = "autoflake>=1.3,<=1.4"
+    default_main = ConsoleScript("autoflake")
+    register_interpreter_constraints = True
+    default_interpreter_constraints = ["CPython==2.7.*", "CPython>=3.4,<3.9"]
```

Comment:
I don't know what the right thing to do here is. This is our first formatter that uses the std-lib AST instead of `typed-ast`, meaning the interp it's run with does matter.

It is not safe to run the formatter in format mode over the same file, which can happen if we partition by interp constraints and there is a Py2 target and Py3 target referring to the same file. While that's valid with a linter, with a formatter, it could produce different results, and then we could have a conflict when trying to merge the digests that will cause Pants to fail. Generally, running a formatter over the same file twice seems undesirable.

Another option is to throw up our hands and always run with a particular interpreter version, based on the subsystem option like this PR does now. That's easiest to implement, but breaks our general support for mixed interpreter constraints: https://blog.pantsbuild.org/python-3-migrations/

It seems like a non-starter to get this formatter to use typed-ast, given that it's trying to emulate Flake8 itself which uses the std lib AST.

A final option is to not land this if we deem it too hard to get working with interpreter constraints and don't want to break that soft guarantee, but that would make me sad and seems undesirable. Two users have reported they would like this feature, one motivated enough to spend his time on the weekend implementing this (thank you Cristian!)

Any other ideas on how to do this, especially that merge conflict risk with using constraints based on the file in question?

--

> Another option is to throw up our hands and always run with a particular interpreter version

I'm leaning towards this, with a warning in the docs and possibly at runtime that this formatter can likely not be used 100% seamlessly with mixed interpreter constraints. Again, I'd be sad if we did not land this due to one workflow not working robustly, as it does work great for most other use cases.

Reply:
Bump @jsirois, thoughts? Does this seem acceptable? 

> Another option is to throw up our hands and always run with a particular interpreter version

>> I'm leaning towards this, with a warning in the docs and possibly at runtime that this formatter can likely not be used 100% seamlessly with mixed interpreter constraints. Again, I'd be sad if we did not land this due to one workflow not working robustly, as it does work great for most other use cases.

Reply:
Isn't it always correct to simply pick a single interpreter - any one will do - from the interpreter constraint range that applies to the file in question? The algorithm we use to pick from that range just needs to be repeatable; so - say picking the max would do.

Reply:
> to the file in question? 

The main question is whether we consult the target's interpreter constraints or not, vs. a single global config option. If we do consult targets, we must afford the possibility of >1 target owning the same file, which means formatting the same file >1 time and possible digest merge conflicts.

I'm proposing that we only look at the global `[autoflake]` subsystem and do not consult individual targets. It will mean Pants does not automatically work for mixed interpreter constraint repos (https://blog.pantsbuild.org/python-3-migrations/), but avoids all the other issues like digest merge conflict errors.

Reply:
So I think what is unsaid in your analysis is you're deeming it too slow to determine the interpreter constraints for a file since to do that you need to discover all its owners; i.e.: walk the full project filesystem. Is that correct?

Reply:
You're suggesting if a Py2 target and Py3 target both own the same file, we would detect this case and go with the Py3 target and ignore the Py2 target? Rather than partitioning by interpreter constraints like we do with Flake8, we would partition by _file_? 

If so, interesting. I think we could do that, although it does introduce entirely new semantics to Pants. When you specify a file arg, that is really a glob for all targets owning that file. Goals like `./pants lint` and `./pants test` will run each relevant target, even if the same file. That gives me caution to add, but maybe there's something there.

Reply:
I have no real knowledge of how things currently work in other goals / rule sets.

From a clean room, the right thing to do is:
1. Determine a valid interpreter to use for a given file in a consistent way.
2. Use it.

Its that simple.

Targets are really just there to trip you up. Yes, we unavoidably currently have them, but I think they should be the absolute last thing on your mind when trying to figure out how something works. Only if they force themselves on you intractably for some reason - perf or legacy tangles - should you pay attention to them.

Reply:
> Its that simple. Targets are really just there to trip you up. 

It would be that simple if we did not allow for a file to have >1 targets, meaning this is legal and even [encouraged in our docs for Py3 migrations](https://www.pantsbuild.org/docs/python-interpreter-compatibility#using-multiple-python-versions-in-the-same-project):

```python
python_tests(name='py2', interpreter_constraints=["==2.7.*"])
python_tests(name='py3', interpreter_constraints=[">=3.6"])
```

> Rather than partitioning by interpreter constraints like we do with Flake8, we would partition by file?

If we do this and only run once-per-file, I realized you'll get confusingly different behavior based on which specs you use:

```bash
./pants fmt path/to/ambiguous.py   # use only one target, e.g. the min interp constraints
./pants fmt path/to/ambiguous.py:py2  # use py2 constraints
./pants fmt path/to/ambiguous.py:py3  # use py3 constraints
```

That's not ideal.

--

So, I think partitioning by file is not realistic with the status quo of interpreter constraints. Meaning, we either give up on autoflake or we do not attempt to robustly handle mixed interpreter constraints. The latter seems far more preferable imo, and users keep asking for autoflake.

@cristianmatache this means keeping what you currently have. We may want to consider changing the default interpreter constraints though to be >=3.5 or even >=3.6...Generally, Pants no longer makes an attempt to support Python 2 Out Of The Box given it's EOL. We want to allow you to configure back in Py2 support, but we would rather optimize for a better experience Out Of The Box with Py3. In this case, by setting the default to 3.5+ or 3.6+, it's more likely that the default config will work with user's repos.

Reply:
Thank you @Eric-Arellano, that sounds like a  very sensible default!

Reply:
@cristianmatache keep on rolling here - my conversation with Eric is just that, you should proceed.

> Determine a valid interpreter to use for a given file in a consistent way.

That means, if a file has more than 1 interpreter constraint, find them all and pick one in a consistent way, lets say lowest.

That's why I brought up speed earlier:

> So I think what is unsaid in your analysis is you're deeming it too slow to determine the interpreter constraints for a file since to do that you need to discover all its owners

Reply:
@jsirois i've been quite busy these days, will get back to it this weekend or next